doc/ipmitool.1
.TH "ipmitool" "1" "" "Duncan Laurie" ""
.SH "NAME"
ipmitool \- utility for controlling IPMI\-enabled devices
.SH "SYNOPSIS"

.BR ipmitool
[ <options> ] <command> [ <sub-commands and sub-options> ]

<options>           := [ <general-options> | <conditional-opts> ]

Any recognized option is accepted. Conditional options may be ignored or it's
usage postponed until shell or exec processes relevant command.

<general\-options>   := [ \-h | \-V | \-v | \-I <interface> | \-H <address> |
                         \-d <N> | \-p <port> | \-c | \-U <username> |
                         \-L <privlvl> | \-l <lun> | \-m <local_address> |
                         \-N <sec> | \-R <count> | <password\-option> |
                         <oem-option> | <bridge-options> ]

<conditional\-opts>  := [ <lan\-options> | <lanplus\-options> |
                         <command\-options> ]
.br
Bridging:
.br
<bridge\-options>    := \-t <target_address> [ \-b <channel> |
                       [ \-T <address> | \-B <channel> ] ]
.br

Options used with \-I lan:
.br
<lan\-options>       := [ \-A <authtype> ]
.br

Options used with \-I lanplus:
.br
<lanplus\-options>   := [ \-C <ciphersuite> | <key\-option> ]
.br

Option groups setting same value:
.br
<key\-option>        := [ \-k <key> | \-K | \-y <hex_key> | \-Y ]
.br
<password\-option>   := [ \-f <password_file> | \-a | \-P <password> | \-E ]
.br
<oem\-option>        := [ \-o <oemtype> | \-g | \-s ]
.br

Options used with specific command <command-options>:
.br
<options\-sdr>       := [ \-S <sdr_cache_file> ]
.br
<options\-sel>       := [ \-O <sel_oem> ]
.br
<options\-sol>       := [ \-e <sol_escape_char> ]


.SH "DESCRIPTION"
This program lets you manage Intelligent Platform Management Interface 
(IPMI) functions of either the local system, via a kernel device driver,
or a remote system, using IPMI v1.5 and IPMI v2.0. These functions include
printing FRU information, LAN configuration, sensor readings, and remote
chassis power control.

IPMI management of a local system interface requires a compatible IPMI
kernel driver to be installed and configured.  On Linux this driver is
called \fIOpenIPMI\fP and it is included in standard distributions.
On Solaris this driver is called \fIBMC\fP and is included in Solaris 10.
Management of a remote station requires the IPMI\-over\-LAN interface to be
enabled and configured.  Depending on the particular requirements of each
system it may be possible to enable the LAN interface using
.BR ipmitool
over the system interface.
.SH "OPTIONS"
.TP 
\fB\-a\fR
Prompt for the remote server password.
.TP 
\fB\-A\fR <\fIauthtype\fP>
Specify an authentication type to use during IPMIv1.5 \fIlan\fP
session activation.  Supported types are NONE, PASSWORD, MD2, MD5, or OEM.
.TP
\fB\-b\fR <\fIchannel\fP>
Set destination channel for bridged request.
.TP
\fB\-B\fR <\fIchannel\fP>
Set transit channel for bridged request (dual bridge).
.TP 
\fB\-b\fR <\fIchannel\fP>
Set destination channel for bridged request.
.TP 
\fB\-B\fR <\fIchannel\fP>
Set transit channel for bridged request. (dual bridge)
.TP 
\fB\-c\fR
Present output in CSV (comma separated variable) format.  
This is not available with all commands.
.TP 
\fB\-C\fR <\fIciphersuite\fP>
The remote server authentication, integrity, and encryption algorithms
to use for IPMIv2.0 \fIlanplus\fP connections.  See table 22\-19 in the
IPMIv2.0 specification.  The default is 3 which specifies RAKP\-HMAC\-SHA1 
authentication, HMAC\-SHA1\-96 integrity, and AES\-CBC\-128 encryption algorithms.
.TP 
\fB\-d \fIN\fP\fR
Use device number N to specify the /dev/ipmiN (or 
/dev/ipmi/N or /dev/ipmidev/N) device to use for in-band 
BMC communication.  Used to target a specific BMC on a 
multi-node, multi-BMC system through the ipmi device 
driver interface.  Default is 0.
.TP 
\fB\-e\fR <\fIsol_escape_char\fP>
Use supplied character for SOL session escape character.  The default
is to use \fI~\fP but this can conflict with ssh sessions.
.TP 
\fB\-E\fR
The remote server password is specified by the environment
variable \fIIPMI_PASSWORD\fP or \fIIPMITOOL_PASSWORD\fP. The \fIIPMITOOL_PASSWORD\fP takes precedence.
.TP 
\fB\-f\fR <\fIpassword_file\fP>
Specifies a file containing the remote server password. If this
option is absent, or if password_file is empty, the password
will default to NULL.
.TP 
\fB\-g\fR
Deprecated. Use: \-o intelplus
.TP 
\fB\-h\fR
Get basic usage help from the command line.
.TP 
\fB\-H\fR <\fIaddress\fP>
Remote server address, can be IP address or hostname.  This 
option is required for \fIlan\fP and \fIlanplus\fP interfaces.
.TP 
\fB\-I\fR <\fIinterface\fP>
Selects IPMI interface to use.  Supported interfaces that are
compiled in are visible in the usage help output.
.TP 
\fB\-k\fR <\fIkey\fP>
Use supplied Kg key for IPMIv2.0 authentication.  The default is not to
use any Kg key.
.TP 
\fB\-K\fR
Read Kg key from IPMI_KGKEY environment variable.
.TP 
\fB\-l\fR <\fIlun\fP>
Set destination lun for raw commands.
.TP 
\fB\-L\fR <\fIprivlvl\fP>
Force session privilege level.  Can be CALLBACK, USER,
OPERATOR, ADMINISTRATOR. Default is ADMINISTRATOR.
This value is ignored and always set to ADMINISTRATOR when
combined with \fI-t target address\fP.
.TP 
\fB\-m\fR <\fIlocal_address\fP>
Set the local IPMB address.  The local address defaults to 0x20
or is auto discovered on PICMG platforms when \-m is not specified.
There should be no need to change the local address for normal operation.
.TP 
\fB\-N\fR <\fIsec\fP>
Specify nr. of seconds between retransmissions of lan/lanplus messages.
Defaults are 2 seconds for lan and 1 second for lanplus interfaces.
Command \fIraw\fP uses fixed value of 15 seconds.
Command \fIsol\fP uses fixed value of 1 second.
.TP 
\fB\-o\fR <\fIoemtype\fP>
Select OEM type to support.  This usually involves minor hacks
in place in the code to work around quirks in various BMCs from
various manufacturers.  Use \fI\-o list\fP to see a list of
current supported OEM types.
.TP 
\fB\-O\fR <\fIsel oem\fP>
Open selected file and read OEM SEL event descriptions to be used
during SEL listings.  See examples in contrib dir for file format.
.TP 
\fB\-p\fR <\fIport\fP>
Remote server UDP port to connect to.  Default is 623.
.TP 
\fB\-P\fR <\fIpassword\fP>
Remote server password is specified on the command line.
If supported it will be obscured in the process list. 
\fBNote!\fR Specifying the password as a command line
option is not recommended.
.TP 
\fB\-R\fR <\fIcount\fP>
Set the number of retries for lan/lanplus interface (default=4).
Command \fIraw\fP uses fixed value of one try (no retries).
Command \fIhpm\fP uses fixed value of 10 retries.
.TP 
\fB\-s\fR
Deprecated. Use: \-o supermicro
.TP 
\fB\-S\fR <\fIsdr_cache_file\fP>
Use local file for remote SDR cache.  Using a local SDR cache
can drastically increase performance for commands that require
knowledge of the entire SDR to perform their function.  Local
SDR cache from a remote system can be created with the
\fIsdr dump\fP command.
.TP 
\fB\-t\fR <\fItarget_address\fP>
Bridge IPMI requests to the remote target address. Default is 32.
The \fI-L privlvl\fP option is always ignored and value set to ADMINISTRATOR.
.TP
\fB\-T\fR <\fIaddress\fP>
Set transit address for bridge request (dual bridge).
.TP 
\fB\-T\fR <\fItransmit_address\fP>
Set transit address for bridge request. (dual bridge)
.TP 
\fB\-U\fR <\fIusername\fP>
Remote server username, default is NULL user.
.TP 
\fB\-v\fR
Increase verbose output level.  This option may be specified
multiple times to increase the level of debug output.  If given
three times you will get hexdumps of all incoming and
outgoing packets. Using it five times provides details
on request and expected reply processing. The \fIhpm\fP commands
\fItargetcap\fP \fIcompprop\fP \fIabort\fP \fIupgstatus\fP
\fIrollback\fP \fIrollbackstatus\fP \fIselftestresult\fP increases
the verbosity level
.TP 
\fB\-V\fR
Display version information.
.TP 
\fB\-y\fR <\fIhex key\fP>
Use supplied Kg key for IPMIv2.0 authentication. The key is expected in
hexadecimal format and can be used to specify keys with non-printable
characters. E.g. '\-k PASSWORD' and '\-y 50415353574F5244' are
equivalent.
The default is not to use any Kg key.
.TP 
\fB\-Y\fR
Prompt for the Kg key for IPMIv2.0 authentication.
.TP
\fB\-z\fR <\fIsize\fP>
Change Size of Communication Channel. (OEM)

.LP 
If no password method is specified then
.BR ipmitool
will prompt the user for a password. If no password is entered at the prompt,
the remote server password will default to NULL.
.SH "SECURITY"
There are several security issues be be considered before enabling the
IPMI LAN interface. A remote station has the ability to control a system's power 
state as well as being able to gather certain platform information. To reduce 
vulnerability it is strongly advised that the IPMI LAN interface only be 
enabled in 'trusted' environments where system security is not an issue or 
where there is a dedicated secure 'management network'.

Further it is strongly advised that you should not enable IPMI for
remote access without setting a password, and that that password should
not be the same as any other password on that system.

When an IPMI password is changed on a remote machine with the IPMIv1.5
\fIlan\fP interface the new password is sent across the network
as clear text.  This could be observed and then used to attack the remote
system.  It is thus recommended that IPMI password management only be done
over IPMIv2.0 \fIlanplus\fP interface or the system interface on the
local station.

For IPMI v1.5, the maximum password length is 16 characters.
Passwords longer than 16 characters will be truncated.

For IPMI v2.0, the maximum password length is 20 characters;
longer passwords are truncated.
.SH "COMMANDS"
.TP 
\fIhelp\fP
This can be used to get command\-line help on 
.BR ipmitool
commands. It may also be placed at the end of commands to get option usage help.

> ipmitool help
.br 
Commands:
        bmc          Deprecated. Use mc
        channel      Configure Management Controller channels
        chassis      Get chassis status and set power state
        dcmi         Data Center Management Interface
        delloem      Manage Dell OEM Extensions.
        echo         Used to echo lines to stdout in scripts
        ekanalyzer   run FRU-Ekeying analyzer using FRU files
        event        Send events to MC
        exec         Run list of commands from file
        firewall     Configure Firmware Firewall
        fru          Print built\-in FRU and scan for FRU locators
        fwum         Update IPMC using Kontron OEM Firmware Update Manager
        gendev       Read/Write Device associated with Generic Device locators sdr
        hpm          Update HPM components using PICMG HPM.1 file
        i2c          Send an I2C Master Write-Read command and print response
        ime          Upgrade/Query Intel ME firmware
        isol         Configure and connect Intel IPMIv1.5 Serial\-over\-LAN
        kontronoem   Manage Kontron OEM Extensions
        lan          Configure LAN Channels
        mc           Management Controller status and global enables
        nm           Node Manager
        pef          Configure Platform Event Filtering (PEF)
        picmg        Run a PICMG/ATA extended command
        power        Shortcut to chassis power commands
        raw          Send a RAW IPMI request and print response
        sdr          Print Sensor Data Repository entries and readings
        sel          Print System Event Log (SEL)
        sensor       Print detailed sensor information
        session      Print session information
        set          Set runtime variable for shell and exec
        shell        Launch interactive IPMI shell
        sol          Configure and connect IPMIv2.0 Serial\-over\-LAN
        spd          Print SPD info from remote I2C device
        sunoem       Manage Sun OEM Extensions
        tsol         Configure and connect Tyan IPMIv1.5 Serial\-over\-LAN
        user         Configure Management Controller users

.TP 
\fIchannel\fP
.RS
.TP 
\fIauthcap\fP <\fBchannel number\fR> <\fBmax priv\fR>

Displays information about the authentication capabilities of
the selected channel at the specified privilege level.
.RS
.TP 
Possible privilege levels are:
.br 
\fI1\fP   Callback level
.br 
\fI2\fP   User level
.br 
\fI3\fP   Operator level
.br 
\fI4\fP   Administrator level
.br 
\fI5\fP   OEM Proprietary level
.br 
\fI15\fP   No access
.RE
.TP 
\fIinfo\fP [\fBchannel number\fR]

Displays  information  about  the selected  channel.  If no channel 
is given it will display information about the currently used channel.
.RS
.PP 
> ipmitool channel info
.br 
Channel 0xf info:
.br 
  Channel Medium Type   : System Interface
.br 
  Channel Protocol Type : KCS
.br 
  Session Support       : session\-less
.br 
  Active Session Count  : 0
.br 
  Protocol Vendor ID    : 7154
.RE
.TP 
\fIgetaccess\fP <\fBchannel number\fR> [<\fBuserid\fR>]
.RS

Configure the given userid as the default on the given channel number.  
When the given channel is subsequently used, the user is identified 
implicitly by the given userid.
.TP 
\fIsetaccess\fP <\fBchannel number\fR> <\fBuserid\fR> [<\fIcallin\fP=\fBon\fR|\fBoff\fR>]
[<\fIipmi\fP=\fBon\fR|\fBoff\fR>] [<\fIlink\fP=\fBon\fR|\fBoff\fR>] [<\fIprivilege\fP=\fBlevel\fR>]
.br 

Configure user access information on the given channel for the given userid.
.TP 
\fIgetciphers\fP <\fIipmi\fP|\fIsol\fP> [<\fBchannel\fR>]
.br 

Displays the list of cipher suites supported for the given
application (ipmi or sol) on the given channel.
.RE
.RE
.TP 
\fIchassis\fP
.RS
.TP 
\fIstatus\fP

Status information related to power, buttons, cooling, drives and faults.
.TP 
\fIpower\fP
.RS
.TP 
\fIstatus\fP
.TP
\fIon\fP
.TP
\fIoff\fP
.TP
\fIcycle\fP
.TP
\fIreset\fP
.TP
\fIdiag\fP
.TP
\fIsoft\fP
.RE
.TP 
\fIidentify\fP [<seconds>|force]

Identify interval.
.br
Default is 15 seconds.
.br
0 - Off
.br
force - To turn on indefinitely
.TP 
\fIpolicy\fP

What to do when power is restored.
.RS
.TP 
\fIlist\fP

Show available options.
.TP 
\fIalways-on\fP
.TP 
\fIprevious\fP
.TP 
\fIalways-off\fP
.RE

.TP 
\fIrestart_cause\fP

Last restart cause.
.TP 
\fIpoh\fP

Get power on hours.
.TP 
\fIbootdev\fP
.RS
.TP 
\fInone\fP

Do not change boot device order.
.TP 
\fIpxe\fP

Force PXE boot.
.TP 
\fIdisk\fP

Force boot from default Hard-drive.
.TP 
\fIsafe\fP

Force boot from default Hard-drive, request Safe Mode.
.TP 
\fIdiag\fP

Force boot from Diagnostic Partition.
.TP 
\fIcdrom\fP

Force boot from CD/DVD.
.TP 
\fIbios\fP

Force boot into BIOS Setup.
.TP 
\fIfloppy\fP

Force boot from Floppy/primary removable media.
.RE

.TP
\fIbootmbox\fP \fIget\fP [text] [block <\fBblock#\fP>]

Read the Boot Initiator Mailbox in hex dump or in text mode.

By default the whole mailbox is read. If block number is specified,
that particular block is read. For block 0 or when the whole
mailbox is read, the Boot Initiator IANA Enterprise Number and
the corresponding enterprise name are printed.

.TP
\fIbootmbox\fP \fIset\fP text [block <\fBblock#\fP>] <\fBIANA_PEN\fP> "<\fBdata_string\fP>"

Write the specified <block> or the entire Boot Initiator Mailbox in text mode.
It is required to specify a decimal IANA Enterprise Number recognized
by the boot initiator on the target system. Refer to your target system
manufacturer for details. The rest of the arguments are a text string.

When single block write is requested, the total length of <data> may not
exceed 13 bytes for block 0, or 16 bytes otherwise.

.TP
\fIbootmbox\fP \fIset\fP [block <\fBblock#\fP>] <\fBIANA_PEN\fP> <\fBdata_byte\fP> [<\fBdata_byte\fP> ...]

Same as above, but the arguments after IANA PEN are separate
data byte values separated by spaces.

.TP 
\fIbootparam\fP \fIget\fP <\fBopt_id\fR> [<\fBopt_param\fR>]

Get value of system boot option number <\fBopt_id\fR>. Some boot
options (e.g. option 7) can also take an optional numeric parameter.

.TP 
\fIbootparam\fP \fIset\fP bootflag <\fBdevice\fR> [options=...]

Set a boot flag. Valid devices are:

.RS
.IP \fIforce_pxe\fP

Force PXE boot
.IP \fIforce_disk\fP

Force boot from default Hard-drive
.IP \fIforce_safe\fP

Force boot from default Hard-drive, request Safe Mode
.IP \fIforce_diag\fP

Force boot from Diagnostic Partition
.IP \fIforce_cdrom\fP

Force boot from CD/DVD
.IP \fIforce_bios\fP

Force boot into BIOS Setup

.PP
Valid options are:

.IP \fIPEF\fP

Clear valid bit on reset/power cycle cause by PEF

.IP \fItimeout\fP

Automatically clear boot flag valid bit on timeout

.IP \fIwatchdog\fP

Clear valid bit on reset/power cycle cause by watchdog

.IP \fIreset\fP

Clear valid bit on push button reset/soft reset

.IP \fIpower\fP

Clear valid bit on power up via power push button or wake event
.RE

.TP 
\fIselftest\fP

Get the chassis self-test results

.RE
.TP 
\fIdcmi\fP
.RS
.TP 
\fIdiscover\fP
.br 

This command is used to discover supported capabilities in DCMI.

.TP 
\fIpower\fP <\fBcommand\fR>
.br 

Platform power limit command options are:

.RS
.TP 
\fIreading\fP
.br 

Get power related readings from the system.
.TP 
\fIget_limit\fP
.br 

Get the configured power limits.
.TP 
\fIset_limit\fP  <\fBparameter\fR> <\fBvalue\fR>
.br 

Set a power limit option.
.br

.RS
.TP
Possible parameters/values are:

.TP
\fIaction\fP <\fBNo Action | Hard Power Off & Log Event to SEL | Log Event to SEL\fR>
.br

Exception Actions are taken as "No Action", "Hard Power Off system and log events to SEL", or "Log event to SEL only".
.TP
\fIlimit\fP <\fBnumber in Watts\fR>
.br

Power Limit Requested in Watts.
.TP
\fIcorrection\fP <\fBnumber in milliseconds\fR>
.br

Correction Time Limit in milliseconds.
.TP
\fIsample\fP <\fBnumber in seconds\fR>
.br

Statistics Sampling period in seconds.

.RE
.TP 
\fIactivate\fP
.br 

Activate the set power limit.
.TP 
\fIdeactivate\fP
.br 

Deactivate the set power limit.
.RE
.TP 
\fIsensors\fP
.br 

Prints the available DCMI sensors.
.TP 
\fIasset_tag\fP
.br 

Prints the platforms asset tag.
.TP 
\fIset_asset_tag\fP  <\fBstring\fR>
.br 

Sets the platforms asset tag
.TP 
\fIget_mc_id_string\fP
.br 

Get management controller identifier string.
.TP 
\fIset_mc_id_string\fP  <\fBstring\fR>
.br 

Set management controller identifier string.  The maximum length is 64 bytes including a null terminator.
.TP 
\fIthermalpolicy\fP  [<\fBget\fR | \fBset\fR>]
.br

Thermal Limit policy get/set.  
.br

.RS
.TP
The commands are:

.TP
\fIGet\fP  <\fBentityID\fR> <\fBinstanceID\fR>
.br 

Get Thermal Limit values.

\fBentityID\fR is the physical entity that a sensor or device is associated with.   
\fBinstanceID\fR is a particular instance of an entity.  Entity Instance can be in one of two ranges, system-relative or device-relative.  For example, a system with four processors could use an Entity Instance value of "0" to identify the first processor.  
.TP
\fISet\fP  <\fBentityID\fR> <\fBinstanceID\fR>
.br 

Set Thermal Limit values.  

\fBentityID\fR is the physical entity that a sensor or device is associated with.   
\fBinstanceID\fR is a particular instance of an entity.  Entity Instance can be in one of two ranges, system-relative or device-relative.  For example, a system with four processors could use an Entity Instance value of "0" to identify the first processor.  
.RE
.TP 
\fIget_temp_reading\fP
.br 

Get Temperature Sensor Readings.
.TP 
\fIget_conf_param\fP
.br 

Get DCMI Configuration Parameters.
.TP 
\fIset_conf_param\fP <\fBparameters\fR>
.br 

Set DCMI Configuration Parameters.  
.RS
.TP
The Configuration Parameters are:
.TP 
\fIactivate_dhcp\fP
.br 

Activate/restart DHCP
.TP 
\fIdhcp_config\fP
.br 

Discover DHCP Configuration.
.TP 
\fIinit\fP 
.br 

Set DHCP Initial timeout interval, in seconds.  The recommended default is four seconds.
.TP 
\fItimeout\fP
.br 

Set DHCP Server contact timeout interval, in seconds.  The recommended default timeout is two minutes.
.TP 
\fIretry\fP
.br 

Set DHCP Server contact retry interval, in seconds.  The recommended default timeout is sixty-four seconds.
.RE
.TP
\fIoob_discover\fP
.br 

Ping/Pong Message for DCMI Discovery.

.RE
.RE
.RE
.TP
\fIdelloem\fP
.RS
.br

The delloem commands provide information on Dell-specific features.
.TP
\fIsetled {b:d.f} {state..}\fP
.RS
.br

Sets the drive backplane LEDs for a device.
.br
{b:d.f} = PCI Address of device (eg. 06:00.0)
.br
{state} = one or more of the following:
.RS
\fIonline | present | hotspare | identify | rebuilding | fault | predict | critical | failed\fP
.br
.RE
.RE

.TP
\fIlcd\fP 
.RS
.br
\fIset {mode}\fP|\fI{lcdqualifier}\fP|\fI{errordisplay}\fP
.RS
.br

Allows you to set the LCD mode and user-defined string.
.RE
.TP
\fIlcd set mode\fP
.RS
.br
\fI{none}\fP|\fI{modelname}\fP|\fI{ipv4address}\fP|\fI{macaddress}\fP|
.br
\fI{systemname}\fP|\fI{servicetag}\fP|\fI{ipv6address}\fP|
.br
\fI{ambienttemp}\fP|\fI{systemwatt}\fP|\fI{assettag}\fP|
.br
\fI{userdefined}<text>\fP
.br

Allows you to set the LCD display mode to any of the preceding parameters.

.RE
.TP
\fIlcd set lcdqualifier\fP
.RS
.br
\fI{watt}\fP|\fI{btuphr}\fP|
.br
\fI{celsius}\fP|\fI{fahrenheit}\fP
.br

Allows you to set the unit for the system ambient temperature mode.

.RE
.TP
\fIlcd set errordisplay\fP
.RS
.br
\fI{sel}\fP|\fI{simple}\fP
.br

Allows you to set the error display.
.RE
.TP
\fIlcd info\fP
.RS
.br

Displays the LCD screen information.
.RE
.TP
\fIlcd set vkvm\fP
.RS
\fI{active}\fP|\fI{inactive}\fP
.br

Allows you to set the vKVM status to active or inactive. When it is active and session is in progress, a message appears on LCD.
.RE
.TP
\fIlcd status\fP
.RS
.br

Displays the LCD status for vKVM display active or inactive and Front Panel access mode (viewandmodify, view-only or disabled).
.RE
.RE
.TP
\fImac\fP
.RS
.br

Displays the information about the system NICs.
.TP
\fImac list\fP
.br

Displays the NIC MAC address and status of all NICs. It also displays the DRAC/iDRAC MAC address.

.TP
\fImac get\fP
.RS
\fI<NIC number>\fP
.br

Displays the selected NICs MAC address and status.
.RE
.RE
.TP
\fIlan\fP
.RS
.br 
Displays the information of Lan.

.TP
\fIlan set\fP
.RS
\fI<Mode>\fP
.br

Sets the NIC selection mode (dedicated, shared with lom1, shared with lom2,shared with lom3,shared with lom4,shared with failover lom1,shared with failover lom2,shared with failover lom3,shared with failover lom4,shared with Failover all loms, shared with Failover None).
.RE
.TP
\fIlan get\fP
.br

Returns the current NIC selection mode (dedicated, shared with lom1, shared with lom2, shared with lom3, shared with lom4,shared with failover lom1, shared with failover lom2,shared with failover lom3,shared with failover lom4,shared with Failover all loms,shared with Failover None).

.TP
\fIlan get active\fP
.br

Returns the current active NIC (dedicated, LOM1, LOM2, LOM3 or LOM4).
.RE
.TP
\fIpowermonitor\fP
.RS
.br

Displays power tracking statistics.

.TP
\fIpowermonitor clear cumulativepower\fP
.RS
.br

Reset cumulative power reading.
.RE
.TP
\fIpowermonitor clear peakpower\fP
.RS
.br

Reset peak power reading.
.RE
.TP
\fIpowermonitor powerconsumption\fP
.RS
\fI<watt>\fP|\fI<btuphr>\fP
.br
Displays the power consumption in watt or btuphr.
.RE
.TP
\fIpowermonitor powerconsumptionhistory\fP
.RS
\fI<watt>\fP|\fI<btuphr>\fP
.br

src/ipmitool.c
/*
 * Copyright (c) 2005 Sun Microsystems, Inc.  All Rights Reserved.
 * Use is subject to license terms.
 */

/*
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 
 * Redistribution of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 * 
 * Redistribution in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 * 
 * Neither the name of Sun Microsystems, Inc. or the names of
 * contributors may be used to endorse or promote products derived
 * from this software without specific prior written permission.
 * 
 * This software is provided "AS IS," without a warranty of any kind.
 * ALL EXPRESS OR IMPLIED CONDITIONS, REPRESENTATIONS AND WARRANTIES,
 * INCLUDING ANY IMPLIED WARRANTY OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE OR NON-INFRINGEMENT, ARE HEREBY EXCLUDED.
 * SUN MICROSYSTEMS, INC. ("SUN") AND ITS LICENSORS SHALL NOT BE LIABLE
 * FOR ANY DAMAGES SUFFERED BY LICENSEE AS A RESULT OF USING, MODIFYING
 * OR DISTRIBUTING THIS SOFTWARE OR ITS DERIVATIVES.  IN NO EVENT WILL
 * SUN OR ITS LICENSORS BE LIABLE FOR ANY LOST REVENUE, PROFIT OR DATA,
 * OR FOR DIRECT, INDIRECT, SPECIAL, CONSEQUENTIAL, INCIDENTAL OR
 * PUNITIVE DAMAGES, HOWEVER CAUSED AND REGARDLESS OF THE THEORY OF
 * LIABILITY, ARISING OUT OF THE USE OF OR INABILITY TO USE THIS SOFTWARE,
 * EVEN IF SUN HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.
 */

#include <ipmitool/ipmi.h>
#include <ipmitool/ipmi_intf.h>
#include <ipmitool/ipmi_main.h>

#include <ipmitool/ipmi_sdr.h>
#include <ipmitool/ipmi_gendev.h>
#include <ipmitool/ipmi_sel.h>
#include <ipmitool/ipmi_fru.h>
#include <ipmitool/ipmi_sol.h>
#include <ipmitool/ipmi_isol.h>
#include <ipmitool/ipmi_tsol.h>
#include <ipmitool/ipmi_lanp.h>
#include <ipmitool/ipmi_chassis.h>
#include <ipmitool/ipmi_mc.h>
#include <ipmitool/ipmi_sensor.h>
#include <ipmitool/ipmi_channel.h>
#include <ipmitool/ipmi_session.h>
#include <ipmitool/ipmi_event.h>
#include <ipmitool/ipmi_user.h>
#include <ipmitool/ipmi_raw.h>
#include <ipmitool/ipmi_pef.h>
#include <ipmitool/ipmi_oem.h>
#include <ipmitool/ipmi_sunoem.h>
#include <ipmitool/ipmi_fwum.h>
#include <ipmitool/ipmi_picmg.h>
#include <ipmitool/ipmi_kontronoem.h>
#include <ipmitool/ipmi_firewall.h>
#include <ipmitool/ipmi_hpmfwupg.h>
#include <ipmitool/ipmi_delloem.h>
#include <ipmitool/ipmi_ekanalyzer.h>
#include <ipmitool/ipmi_ime.h>
#include <ipmitool/ipmi_dcmi.h>
#include <ipmitool/ipmi_vita.h>
#include <ipmitool/ipmi_quantaoem.h>

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#ifdef HAVE_READLINE
extern int ipmi_shell_main(struct ipmi_intf * intf, int argc, char ** argv);
#endif
extern int ipmi_echo_main(struct ipmi_intf * intf, int argc, char ** argv);
extern int ipmi_set_main(struct ipmi_intf * intf, int argc, char ** argv);
extern int ipmi_exec_main(struct ipmi_intf * intf, int argc, char ** argv);
extern int ipmi_lan6_main(struct ipmi_intf *intf, int argc, char **argv);


int csv_output = 0;
int time_in_utc = 0;
int verbose = 0;

struct ipmi_cmd ipmitool_cmd_list[] = {
	{ ipmi_raw_main,     "raw",     "Send a RAW IPMI request and print response" },
	{ ipmi_rawi2c_main,  "i2c",     "Send an I2C Master Write-Read command and print response" },
	{ ipmi_rawspd_main,  "spd",     "Print SPD info from remote I2C device" },
	{ ipmi_lanp_main,    "lan",     "Configure LAN Channels" },
	{ ipmi_chassis_main, "chassis", "Get chassis status and set power state" },
	{ ipmi_power_main,   "power",   "Shortcut to chassis power commands" },
	{ ipmi_event_main,   "event",   "Send pre-defined events to MC" },
	{ ipmi_mc_main,      "mc",      "Management Controller status and global enables" },
	{ ipmi_mc_main,      "bmc",     NULL },	/* for backwards compatibility */
	{ ipmi_sdr_main,     "sdr",     "Print Sensor Data Repository entries and readings" },
	{ ipmi_sensor_main,  "sensor",  "Print detailed sensor information" },
	{ ipmi_fru_main,     "fru",     "Print built-in FRU and scan SDR for FRU locators" },
	{ ipmi_gendev_main,  "gendev",  "Read/Write Device associated with Generic Device locators sdr" },
	{ ipmi_sel_main,     "sel",     "Print System Event Log (SEL)" },
	{ ipmi_pef_main,     "pef",     "Configure Platform Event Filtering (PEF)" },
	{ ipmi_sol_main,     "sol",     "Configure and connect IPMIv2.0 Serial-over-LAN" },
	{ ipmi_tsol_main,    "tsol",    "Configure and connect with Tyan IPMIv1.5 Serial-over-LAN" },
	{ ipmi_isol_main,    "isol",    "Configure IPMIv1.5 Serial-over-LAN" },
	{ ipmi_user_main,    "user",    "Configure Management Controller users" },
	{ ipmi_channel_main, "channel", "Configure Management Controller channels" },
	{ ipmi_session_main, "session", "Print session information" },
	{ ipmi_dcmi_main,    "dcmi",    "Data Center Management Interface"},
	{ ipmi_nm_main,      "nm",      "Node Manager Interface"},
	{ ipmi_sunoem_main,  "sunoem",  "OEM Commands for Sun servers" },
	{ ipmi_kontronoem_main, "kontronoem", "OEM Commands for Kontron devices"},
	{ ipmi_picmg_main,   "picmg",   "Run a PICMG/ATCA extended cmd"},
	{ ipmi_fwum_main,    "fwum",	"Update IPMC using Kontron OEM Firmware Update Manager" },
	{ ipmi_firewall_main,"firewall","Configure Firmware Firewall" },
	{ ipmi_delloem_main, "delloem", "OEM Commands for Dell systems" },
#ifdef HAVE_READLINE
	{ ipmi_shell_main,   "shell",   "Launch interactive IPMI shell" },
#endif
	{ ipmi_exec_main,    "exec",    "Run list of commands from file" },
	{ ipmi_set_main,     "set",     "Set runtime variable for shell and exec" },
	{ ipmi_echo_main,    "echo",    NULL }, /* for echoing lines to stdout in scripts */
	{ ipmi_hpmfwupg_main,"hpm", "Update HPM components using PICMG HPM.1 file"},
	{ ipmi_ekanalyzer_main,"ekanalyzer", "run FRU-Ekeying analyzer using FRU files"},
	{ ipmi_ime_main,          "ime", "Update Intel Manageability Engine Firmware"},
	{ ipmi_vita_main,   "vita",   "Run a VITA 46.11 extended cmd"},
	{ ipmi_lan6_main,   "lan6",   "Configure IPv6 LAN Channels"},
	{ NULL },
};

int
main(int argc, char ** argv)
{
	int rc;

	rc = ipmi_main(argc, argv, ipmitool_cmd_list, NULL);

	if (rc < 0)
		exit(EXIT_FAILURE);
	else
		exit(EXIT_SUCCESS);
}
